File: examples/python/aws_secrets.py
"""
AWS Secrets Manager Integration Module
=======================================
This module demonstrates how to securely fetch secrets from AWS Secrets Manager
using boto3. All configuration is bootstrapped from environment variables —
never from hardcoded values — so the same code works across dev, staging, and
production with zero changes.

Security rationale
------------------
* Credentials (access key, secret key, session token) come from the
  environment or an IAM instance/profile role.  Nothing is baked into
  source control.
* boto3 resolves credentials transparently via its credential provider
  chain (env vars → ~/.aws/credentials → IAM role).  This means we
  never see or handle raw AWS credentials in application code.
* Secret values are never logged; we log only the *name* of the secret
  that was fetched and whether the operation succeeded.
* SecretString payloads are expected to be JSON, giving us type-safe
  access to structured configuration (e.g. DB host, port, username).
"""

import json
import logging
import os
from typing import Any

import boto3
from botocore.exceptions import ClientError

# ---------------------------------------------------------------------------
# Structured logging — we use JSON-like formatting so log aggregators
# (Datadog, CloudWatch, ELK) can parse fields automatically.
# IMPORTANT: we never include secret *values* in log output.
# ---------------------------------------------------------------------------
logger = logging.getLogger("aws_secrets")
logger.setLevel(logging.INFO)


def _log_event(event: str, secret_name: str | None = None, **extra: Any) -> None:
    """Emit a structured log entry without exposing secret values."""
    payload = {"event": event, "secret_name": secret_name, **extra}
    # Mask any key that looks like it could contain a value
    for k, v in payload.items():
        if any(s in k.lower() for s in ("key", "token", "secret", "password")):
            payload[k] = "***REDACTED***"
    logger.info(payload)


class AWSSecretsManager:
    """High-level wrapper around AWS Secrets Manager.

    Each public method fetches a *specific* secret by name and returns a
    typed, validated dictionary.  Keeping the mapping from secret name
    to expected shape in one place makes it easy to audit what secrets
    the application depends on.
    """

    # Default secret names — override via constructor or env vars so that
    # different environments (dev/stage/prod) can point at different secrets
    # without code changes.
    DEFAULT_OPENAI_SECRET = "prod/openai/config"
    DEFAULT_DB_SECRET = "prod/database/credentials"
    DEFAULT_JWT_SECRET = "prod/jwt/config"

    def __init__(
        self,
        region_name: str | None = None,
        openai_secret_name: str | None = None,
        db_secret_name: str | None = None,
        jwt_secret_name: str | None = None,
    ) -> None:
        # ------------------------------------------------------------------
        # Environment-variable bootstrapping:
        # We read AWS_REGION, AWS_ACCESS_KEY_ID etc. from the environment.
        # boto3's credential chain handles the rest automatically, so we
        # only pass region_name explicitly when the caller provides one
        # (or when AWS_REGION is set).
        # ------------------------------------------------------------------
        self._region = region_name or os.getenv("AWS_REGION", "us-east-1")

        # Log only the region, never the keys themselves
        _log_event("aws_client_init", extra={"region": self._region})

        # Create a boto3 session — this is the recommended way to configure
        # the SDK because it isolates our client from global state.
        session = boto3.session.Session()
        self._client = session.client(
            service_name="secretsmanager",
            region_name=self._region,
        )

        # Resolve secret names: constructor arg > env var > class default
        self._openai_secret = (
            openai_secret_name
            or os.getenv("AWS_SECRET_OPENAI", self.DEFAULT_OPENAI_SECRET)
        )
        self._db_secret = (
            db_secret_name
            or os.getenv("AWS_SECRET_DB", self.DEFAULT_DB_SECRET)
        )
        self._jwt_secret = (
            jwt_secret_name
            or os.getenv("AWS_SECRET_JWT", self.DEFAULT_JWT_SECRET)
        )

    # ------------------------------------------------------------------
    # Core fetch method
    # ------------------------------------------------------------------
    def get_secret(self, secret_name: str) -> dict[str, Any]:
        """Fetch a single secret by name and return it as a parsed dict.

        The secret value stored in AWS Secrets Manager is expected to be a
        JSON string (the AWS console encourages this).  If parsing fails we
        raise a ValueError rather than returning opaque data.

        Raises:
            ClientError: on AWS API failures (access denied, not found, …)
            ValueError: when the secret payload is not valid JSON.
        """
        _log_event("fetching_secret", secret_name=secret_name)
        try:
            response = self._client.get_secret_value(SecretId=secret_name)
        except ClientError as exc:
            # Log the *error code* (e.g. ResourceNotFoundException) but
            # never the raw response body which could leak information.
            error_code = exc.response.get("Error", {}).get("Code", "Unknown")
            _log_event(
                "fetch_secret_failed",
                secret_name=secret_name,
                error_code=error_code,
            )
            raise

        # Secrets Manager may return SecretString or SecretBinary.
        if "SecretString" in response:
            secret_payload = response["SecretString"]
        elif "SecretBinary" in response:
            # Binary secrets are base64-encoded bytes — decode if needed.
            secret_payload = response["SecretBinary"].decode("utf-8")
        else:
            raise ValueError(f"Secret '{secret_name}' has no string or binary payload")

        try:
            parsed: dict[str, Any] = json.loads(secret_payload)
        except json.JSONDecodeError as exc:
            raise ValueError(
                f"Secret '{secret_name}' is not valid JSON"
            ) from exc

        _log_event("secret_fetched", secret_name=secret_name)
        return parsed

    # ------------------------------------------------------------------
    # Typed convenience methods
    # ------------------------------------------------------------------
    def get_database_credentials(self) -> dict[str, Any]:
        """Return database connection parameters from Secrets Manager.

        Expected secret shape:
        {
            "engine": "postgresql",
            "host": "db.example.com",
            "port": 5432,
            "username": "app_user",
            "password": "<sensitive>",
            "dbname": "app_db",
            "ssl_mode": "require"
        }

        SECURITY NOTE: The password is present in the returned dict.
        The caller must ensure it is never logged or serialized to disk.
        """
        creds = self.get_secret(self._db_secret)
        required_fields = {"host", "username", "password", "dbname"}
        missing = required_fields - set(creds.keys())
        if missing:
            raise ValueError(f"DB secret missing fields: {missing}")
        return creds

    def get_jwt_config(self) -> dict[str, Any]:
        """Return JWT signing configuration from Secrets Manager.

        Expected secret shape:
        {
            "algorithm": "HS256",
            "secret_key": "<signing-key>",
            "access_token_ttl_minutes": 15,
            "refresh_token_ttl_days": 7
        }
        """
        config = self.get_secret(self._jwt_secret)
        required_fields = {"algorithm", "secret_key"}
        missing = required_fields - set(config.keys())
        if missing:
            raise ValueError(f"JWT secret missing fields: {missing}")
        return config

    def get_openai_config(self) -> dict[str, Any]:
        """Return OpenAI API configuration from Secrets Manager.

        Expected secret shape:
        {
            "api_key": "sk-...",
            "organization_id": "org-...",
            "default_model": "gpt-4",
            "max_tokens": 2048
        }
        """
        config = self.get_secret(self._openai_secret)
        if "api_key" not in config:
            raise ValueError("OpenAI secret missing 'api_key' field")
        return config

    # ------------------------------------------------------------------
    # Secret rotation awareness
    # ------------------------------------------------------------------
    def describe_rotation(self, secret_name: str) -> dict[str, Any]:
        """Check whether automatic rotation is configured for a secret.

        Rotation is critical for long-lived secrets like database passwords.
        AWS Secrets Manager can invoke a Lambda function on a schedule to
        rotate the value transparently.  This method lets the application
        verify at startup that rotation is enabled.
        """
        try:
            response = self._client.describe_secret(SecretId=secret_name)
        except ClientError as exc:
            error_code = exc.response.get("Error", {}).get("Code", "Unknown")
            _log_event("describe_rotation_failed", secret_name=secret_name, error_code=error_code)
            raise

        rotation_enabled = response.get("RotationEnabled", False)
        rotation_lambda = response.get("RotationLambdaARN", "")

        _log_event(
            "rotation_status",
            secret_name=secret_name,
            rotation_enabled=rotation_enabled,
            # Log only the Lambda ARN — not any secret material
            rotation_lambda_arn=rotation_lambda or "none",
        )
        return {
            "secret_name": secret_name,
            "rotation_enabled": rotation_enabled,
            "rotation_lambda_arn": rotation_lambda,
        }


# ---------------------------------------------------------------------------
# Quick-start demonstration (run as a script)
# ---------------------------------------------------------------------------
if __name__ == "__main__":
    logging.basicConfig(level=logging.INFO, format="%(message)s")

    mgr = AWSSecretsManager()

    # Demonstrate fetching each type of secret
    try:
        db_creds = mgr.get_database_credentials()
        print(f"DB host: {db_creds['host']}")  # safe — not a secret
    except Exception as exc:
        print(f"Could not fetch DB creds (expected in local dev): {exc}")

    try:
        jwt_cfg = mgr.get_jwt_config()
        print(f"JWT algorithm: {jwt_cfg['algorithm']}")
    except Exception as exc:
        print(f"Could not fetch JWT config (expected in local dev): {exc}")

    try:
        rotation = mgr.describe_rotation(mgr._db_secret)
        print(f"DB rotation enabled: {rotation['rotation_enabled']}")
    except Exception as exc:
        print(f"Could not describe rotation (expected in local dev): {exc}")

File: examples/nodejs/aws-secrets.js
/**
 * AWS Secrets Manager Integration
 *
 * WHY THIS APPROACH OVER .env FILES:
 * ─────────────────────────────────
 * .env files are risky because:
 *   1. They are typically committed to version control (even accidentally)
 *   2. They sit on disk in plaintext — any process/user with read access can leak them
 *   3. There's no built-in rotation — a compromised secret stays compromised
 *   4. They don't scale across multiple environments or services
 *   5. They lack audit trails — you can't track who accessed what and when
 *
 * AWS Secrets Manager solves these problems by:
 *   - Storing secrets encrypted at rest (AES-256)
 *   - Providing automatic secret rotation via Lambda functions
 *   - Offering fine-grained IAM policies for access control
 *   - Emitting CloudTrail logs for every access event
 *   - Integrating with ECS, EKS, Lambda, EC2 via IAM roles (no credentials needed)
 */

const { SecretsManagerClient, GetSecretValueCommand } = require("@aws-sdk/client-secrets-manager");
const winston = require("winston");

// ── Logger Configuration ──────────────────────────────────────────────────
// CRITICAL: Never log secret values. Even partial values can be used in
// credential stuffing attacks. Always use the [REDACTED] placeholder.
const logger = winston.createLogger({
  level: process.env.LOG_LEVEL || "info",
  format: winston.format.combine(
    winston.format.timestamp(),
    winston.format.errors({ stack: true }),
    winston.format.json()
  ),
  transports: [new winston.transports.Console()],
});

// ── AWS SDK Client Initialization ──────────────────────────────────────────
// The region is read from AWS_REGION (or AWS_DEFAULT_REGION) environment variable.
// This is the standard AWS convention — never hardcode a region because
// deployments may span multiple regions for disaster recovery.
let secretsClient;

/**
 * Initializes the AWS Secrets Manager client.
 * Called lazily so the module can be imported without requiring AWS credentials
 * at import time (useful for testing or non-AWS environments).
 */
function getClient() {
  if (!secretsClient) {
    const region = process.env.AWS_REGION || process.env.AWS_DEFAULT_REGION;
    if (!region) {
      throw new Error(
        "AWS_REGION environment variable is required. " +
        "Set it to your deployment region (e.g., 'us-east-1')."
      );
    }
    secretsClient = new SecretsManagerClient({ region });
    logger.info("AWS Secrets Manager client initialized", { region });
  }
  return secretsClient;
}

/**
 * Fetches a secret by its name/ARN from AWS Secrets Manager.
 *
 * SECURITY NOTE: We parse the SecretString as JSON if possible. Many teams
 * store multiple key-value pairs in a single secret (e.g., both username
 * and password). The returned object gives callers structured access without
 * exposing the raw string.
 *
 * @param {string} secretName - The name or ARN of the secret in AWS.
 * @returns {Promise<object|string>} Parsed JSON object or raw string.
 */
async function getSecret(secretName) {
  const client = getClient();

  logger.info("Fetching secret from AWS Secrets Manager", {
    secretName,
    // NEVER log the secret value or even a partial value
  });

  try {
    const command = new GetSecretValueCommand({ SecretId: secretName });
    const response = await client.send(command);

    // Secrets Manager may return the value as a binary blob (SecretBinary)
    // or a UTF-8 string (SecretString). We handle both.
    if (response.SecretString) {
      try {
        // Most secrets are stored as JSON strings for structured access
        const parsed = JSON.parse(response.SecretString);
        logger.info("Secret retrieved successfully", {
          secretName,
          keyCount: Object.keys(parsed).length,
          value: "[REDACTED]",
        });
        return parsed;
      } catch {
        // Not JSON — return the raw string (e.g., a plaintext API key)
        logger.info("Secret retrieved (raw string)", {
          secretName,
          value: "[REDACTED]",
        });
        return response.SecretString;
      }
    }

    if (response.SecretBinary) {
      // Binary secrets (e.g., certificates) are returned as Base64-encoded Buffers
      logger.info("Secret retrieved (binary)", {
        secretName,
        value: "[REDACTED]",
      });
      return Buffer.from(response.SecretBinary, "base64");
    }

    throw new Error("Secret returned neither SecretString nor SecretBinary");
  } catch (error) {
    // Distinguish between "secret doesn't exist" and "access denied" errors
    if (error.name === "ResourceNotFoundException") {
      logger.error("Secret not found in AWS Secrets Manager", { secretName });
      throw new Error(`Secret '${secretName}' does not exist or you lack access`);
    }
    if (error.name === "AccessDeniedException") {
      // SECURITY: Log the denial but don't expose IAM details to callers
      logger.error("Access denied to secret — check IAM permissions", {
        secretName,
        errorCode: error.Code,
      });
      throw new Error("Permission denied when accessing secret");
    }
    logger.error("Failed to fetch secret from AWS", {
      secretName,
      error: error.message,
    });
    throw error;
  }
}

/**
 * Fetches database credentials from a named secret.
 *
 * Expected secret structure (JSON):
 *   { "username": "app_user", "password": "s3cret!", "engine": "postgres",
 *     "host": "db.example.com", "port": 5432, "dbname": "myapp" }
 *
 * ROTATION CONCEPT: AWS Secrets Manager supports automatic rotation via
 * Lambda functions. When a database password rotates, the Lambda:
 *   1. Generates a new password
 *   2. Updates the secret in Secrets Manager
 *   3. Updates the password in the database
 *   4. (Optionally) invalidates existing connections
 *
 * Our application simply re-fetches the secret to get the new credentials.
 * No code change or redeployment is needed.
 *
 * @param {string} secretName - Secret name (default: "prod/database/credentials")
 * @returns {Promise<object>} Database connection parameters.
 */
async function getDatabaseCredentials(secretName = "prod/database/credentials") {
  const secret = await getSecret(secretName);

  // Validate the secret contains expected database fields
  const required = ["username", "password", "host"];
  for (const field of required) {
    if (!secret[field]) {
      logger.error("Database secret missing required field", {
        secretName,
        missingField: field,
      });
      throw new Error(`Database secret '${secretName}' is missing field '${field}'`);
    }
  }

  logger.info("Database credentials retrieved", {
    secretName,
    host: secret.host,
    port: secret.port || 5432,
    engine: secret.engine || "unknown",
    username: secret.username,
    password: "[REDACTED]",
  });

  return {
    host: secret.host,
    port: secret.port || 5432,
    database: secret.dbname || secret.database,
    username: secret.username,
    password: secret.password, // Passed programmatically, never logged
    ssl: secret.ssl !== false, // Default to SSL enabled
    connectionTimeout: secret.connectionTimeout || 5000,
  };
}

/**
 * Fetches JWT signing configuration from Secrets Manager.
 *
 * Expected secret structure (JSON):
 *   { "algorithm": "HS256", "secretKey": "your-256-bit-secret",
 *     "issuer": "myapp", "expiresIn": "15m" }
 *
 * @param {string} secretName - Secret name (default: "prod/jwt/config")
 * @returns {Promise<object>} JWT configuration.
 */
async function getJwtConfig(secretName = "prod/jwt/config") {
  const secret = await getSecret(secretName);

  logger.info("JWT configuration retrieved", {
    secretName,
    algorithm: secret.algorithm || "HS256",
    issuer: secret.issuer || "[REDACTED]",
    secretKey: "[REDACTED]",
  });

  return {
    algorithm: secret.algorithm || "HS256",
    secretKey: secret.secretKey,
    issuer: secret.issuer || "secrets-management-app",
    audience: secret.audience,
    expiresIn: secret.expiresIn || "15m",
  };
}

/**
 * Fetches OpenAI API configuration from Secrets Manager.
 *
 * This demonstrates fetching third-party API credentials. Storing these
 * in Secrets Manager (not .env) ensures they can be rotated independently
 * and audited. If an API key is compromised, you rotate it in one place
 * and all services pick up the new key on next fetch.
 *
 * @param {string} secretName - Secret name (default: "prod/openai/config")
 * @returns {Promise<object>} OpenAI configuration.
 */
async function getOpenaiConfig(secretName = "prod/openai/config") {
  const secret = await getSecret(secretName);

  if (!secret.apiKey) {
    logger.error("OpenAI secret missing apiKey field", { secretName });
    throw new Error(`Secret '${secretName}' must contain an 'apiKey' field`);
  }

  logger.info("OpenAI configuration retrieved", {
    secretName,
    organization: secret.organization || "default",
    apiKey: "[REDACTED]",
    apiKeyPrefix: secret.apiKey ? `${secret.apiKey.substring(0, 7)}...` : "[REDACTED]",
    // Logging only the prefix (e.g., "sk-proj...") is generally safe
    // and helps debug which key is in use without exposing the full value.
  });

  return {
    apiKey: secret.apiKey,
    organization: secret.organization,
    model: secret.model || "gpt-4",
    maxTokens: secret.maxTokens || 2048,
  };
}

module.exports = {
  getSecret,
  getDatabaseCredentials,
  getJwtConfig,
  getOpenaiConfig,
};